Mathlib/NumberTheory/NumberField/CanonicalEmbedding/ConvexBody.lean

```diff
@@ -538,38 +538,41 @@ theorem exists_ne_zero_mem_ideal_lt' (w₀ : {w : InfinitePlace K // IsComplex w
 
 /-- A version of `exists_ne_zero_mem_ideal_lt` for the ring of integers of `K`. -/
 theorem exists_ne_zero_mem_ringOfIntegers_lt (h : minkowskiBound K ↑1 < volume (convexBodyLT K f)) :
-    ∃ a ∈ 𝓞 K, a ≠ 0 ∧ ∀ w : InfinitePlace K, w a < f w := by
+    ∃ a : 𝓞 K, a ≠ 0 ∧ ∀ w : InfinitePlace K, w (algebraMap _ _ a) < f w := by
   obtain ⟨_, h_mem, h_nz, h_bd⟩ := exists_ne_zero_mem_ideal_lt K ↑1 h
-  obtain ⟨⟨a, ha⟩, rfl⟩ := (FractionalIdeal.mem_one_iff _).mp h_mem
-  exact ⟨a, ha, h_nz, h_bd⟩
+  obtain ⟨a, rfl⟩ := (FractionalIdeal.mem_one_iff _).mp h_mem
+  exact ⟨a, mt RingOfIntegers.coe_eq_zero_iff.mpr h_nz, h_bd⟩
 
 /-- A version of `exists_ne_zero_mem_ideal_lt'` for the ring of integers of `K`. -/
 theorem exists_ne_zero_mem_ringOfIntegers_lt' (w₀ : {w : InfinitePlace K // IsComplex w})
     (h : minkowskiBound K ↑1 < volume (convexBodyLT' K f w₀)) :
-    ∃ a ∈ 𝓞 K, a ≠ 0 ∧ (∀ w : InfinitePlace K, w ≠ w₀ → w a < f w) ∧
-      |(w₀.val.embedding a).re| < 1 ∧ |(w₀.val.embedding a).im| < (f w₀ : ℝ) ^ 2 := by
+    ∃ a : 𝓞 K, a ≠ 0 ∧ (∀ w : InfinitePlace K, w ≠ w₀ → w (algebraMap _ _ a) < f w) ∧
+      |(w₀.val.embedding (algebraMap _ _ a)).re| < 1 ∧
+      |(w₀.val.embedding (algebraMap _ _ a)).im| < (f w₀ : ℝ) ^ 2 := by
```

Comment:
How about adding a coercion from O_K to K? I remember @Vierkantor once explaining to me why it was not possible to make algebraMap a coercion (something to do with head symbols, but I have forgotten the details, probably this means I didn't understand them properly). But is it possible here?

Reply:
I didn't define the coercion because I thought it would require a lot more work to set up the API: that it preserves the ring operations etc. With `algebraMap` we get that for free, in exchange for a lot more writing.

The reason we can't say `instance : CoeHead (𝓞 K) K := ⟨algebraMap _ _⟩` is that all the metaprogramming needs a `@[coe]` attribute on the expression to identify a coercion. Specifically on the head symbol. In this case, because `algebraMap` is a bundled function, the head symbol is different from `algebraMap` in this case. We'd need to define a synonym for this map that is not bundled and we can give that the `@[coe]` attribute.

Following what @erdOne suggests, the coercion be defined something as:
```lean
@[coe]
abbrev val (x : 𝓞 K) : K := algebraMap _ _ x

/-- This instance has to be `CoeHead` because we only want to apply it from `𝓞 K` to `K`. -/
instance : CoeHead (𝓞 K) K := ⟨val⟩
```
Although I can't say for sure if making `val` an abbrev would work, perhaps it has to be a `def`.